miio/gateway.py

```diff
@@ -126,58 +125,64 @@ def light(self) -> "GatewayLight":
         """Return light control interface."""
         return self._light
 
-    @command()
+    @property
     def devices(self):
-        """Return list of devices."""
+        """Return a list of the already discovered devices."""
+        return self._devices
+
+    @command()
+    def discover_devices(self):
+        """Discovers SubDevices and returns a list of the discovered devices."""
         # from https://github.com/aholstenson/miio/issues/26
-        devices_raw = self.send("get_device_prop", ["lumi.0", "device_list"])
-        devices = [
-            SubDevice(self, *devices_raw[x : x + 5])  # noqa: E203
-            for x in range(0, len(devices_raw), 5)
-        ]
+        device_type_mapping = {
+            "AqaraRelayTwoChannels": AqaraRelayTwoChannels,
+            "Plug": AqaraPlug,
+            "SensorHT": SensorHT,
+            "AqaraHT": AqaraHT,
+            "AqaraMagnet": AqaraMagnet,
+        }
+        devices_raw = self.get_prop("device_list")
+        self._devices = []
+
+        for x in range(0, len(devices_raw), 5):
```

Comment:
Related to the dev_type number conversation below, I think the first line in this loop should be:
```dev_type_id = devices_raw[x + 1]```

Maybe even extract rest of the known elements here, too, instead of passing `*devices_raw[x : x + 5]` later on? This would make the code much more understandable.

Reply:
sounds good, I know the unknown3 == firmware version number of the subdevice

Reply:
at least `self.get_subdevice_prop("fw_ver").pop()` results in 3x '3' and 2x '9' for my 5 devices and those numbers match the unknown3 parameter

Reply:
I have implemented a dataclass and made it a bit nicer

Reply:
Great! :100: